File: python-code/updater.py
import csv
import datetime
import json
import os
import re
import tempfile
import time
from typing import List, Dict, Any, Tuple

from dotenv import load_dotenv
from pymongo.results import InsertManyResult

from db import MongoDBConnection
from download_file import download_file
from utils.calculate_checksum import calculate_checksum
from utils.create_unique_key import create_unique_key
from utils.extract_zip_file import extract_zip_file
from utils.redis_cache import get_cached_checksum, cache_checksum

load_dotenv()


def read_csv_data(file_path: str) -> List[Dict[str, Any]]:
    csv_data = []
    with open(file_path, "r", encoding="utf-8") as csv_file:
        csv_reader = csv.DictReader(csv_file)
        for row in csv_reader:
            csv_data.append(process_csv_row(row))
    return csv_data


def process_csv_row(row) -> Dict[str, Any]:
    if "make" in row:
        row["make"] = row["make"].replace(".", "")
    # Convert string values to numbers if possible
    for key, value in row.items():
        if re.match(r"\b\d+(?:,\d+)?\b", value):
            row[key] = 0 if value == "" else value
            try:
                value = str(int(value.replace(",", "")))
                row[key] = float(value) if "." in value else int(value)
            except ValueError:
                pass
    return row


async def updater(
    collection_name: str, zip_file_name: str, zip_url: str, key_fields: List[str]
) -> Tuple[InsertManyResult | None, str]:
    db = MongoDBConnection().database
    collection = db[collection_name]

    try:
        with tempfile.TemporaryDirectory() as temp_dir:
            zip_file_path = os.path.join(temp_dir, zip_file_name)
            download_file(zip_url, zip_file_path)

            extracted_file_name = extract_zip_file(zip_file_path, temp_dir)
            destination_path = os.path.join(temp_dir, extracted_file_name)
            print(f"Destination path: {destination_path}")

            checksum = calculate_checksum(destination_path)
            print(f"Checksum (SHA-256) of extracted file: {checksum}")

            cached_checksum = await get_cached_checksum(extracted_file_name)
            if cached_checksum is None:
                print("No cached checksum found. This might be the first run.")
                await cache_checksum(extracted_file_name, checksum)
                print(
                    f"Checksum for {extracted_file_name} cached. Checksum: {checksum}"
                )
            elif cached_checksum == checksum:
                return (
                    None,
                    f"File have not been changed since last update. Checksum: {checksum}",
                )

            await cache_checksum(extracted_file_name, checksum)
            print("Checksum has been changed.")

            csv_data: List[Dict[str, Any]] = read_csv_data(destination_path)

            existing_data_map = {
                create_unique_key(item, key_fields): item
                for item in collection.find({})
            }
            new_data_to_insert = [
                item
                for item in csv_data
                if create_unique_key(item, key_fields) not in existing_data_map
            ]

            if new_data_to_insert:
                start = time.time()
                result = collection.insert_many(new_data_to_insert)
                end = time.time()
                message = f"{len(result.inserted_ids)} document(s) inserted in {round((end - start) * 1000)}ms"
                return result, message
            else:
                message = "No new data to insert. The provided data matches the existing records."
                return None, message

    except Exception as error:
        raise Exception(f"An error has occurred: {error}")
    finally:
        db.client.close()


async def main(
    collection_name: str, zip_file_name: str, zip_url: str, key_fields: List[str]
) -> Dict[str, Any]:
    timestamp = datetime.datetime.now().isoformat()

    try:
        result, message = await updater(
            collection_name,
            zip_file_name,
            zip_url,
            key_fields,
        )
        inserted_count = len(result.inserted_ids) if result else 0
    except Exception as e:
        inserted_count = 0
        message = str(e)

    response = {
        "collection": collection_name,
        "inserted_count": inserted_count,
        "message": message,
        "timestamp": timestamp,
    }

    print(json.dumps(response, indent=4))
    return response

File: src/lib/updater.ts
import { downloadFile } from "../utils/downloadFile";
import { AWS_LAMBDA_TEMP_DIR } from "../config";
import path from "node:path";
import { calculateChecksum } from "../utils/calculateChecksum";
import { cacheChecksum, getCachedChecksum } from "../utils/redisCache";
import { processCSV } from "../utils/processCSV";
import { db } from "../db";
import { PgTable } from "drizzle-orm/pg-core";
import { createUniqueKey } from "../utils/createUniqueKey";

export interface UpdaterConfig<T extends PgTable> {
  table: T;
  zipFileName: string;
  zipUrl: string;
  keyFields: string[];
}

export interface UpdaterResult {
  recordsProcessed: number;
  message: string;
  timestamp: string;
}

const BATCH_SIZE = 5000;

const downloadAndPrepareFile = async (zipUrl: string): Promise<string> => {
  const extractedFileName = await downloadFile(zipUrl);
  const destinationPath = path.join(AWS_LAMBDA_TEMP_DIR, extractedFileName);
  console.log("Destination path:", destinationPath);
  return destinationPath;
};

const validateFileChecksum = async (
  extractedFileName: string,
  destinationPath: string,
): Promise<{ checksum: string; isNewOrChanged: boolean }> => {
  const checksum = await calculateChecksum(destinationPath);
  console.log("Checksum:", checksum);

  const cachedChecksum = await getCachedChecksum(extractedFileName);
  console.log("Cached checksum:", cachedChecksum);

  if (!cachedChecksum) {
    console.log(`No cached checksum found. This might be the first run.`);
    await cacheChecksum(extractedFileName, checksum);
    return { checksum, isNewOrChanged: true };
  }

  const isNewOrChanged = cachedChecksum !== checksum;
  if (!isNewOrChanged) {
    console.log(
      `File has not been changed since the last update. Checksum: ${checksum}`,
    );
  }

  return { checksum, isNewOrChanged };
};

const findNewRecords = async <T extends PgTable>(
  table: T,
  processedData: Record<string, any>[],
  keyFields: string[],
): Promise<Record<string, any>[]> => {
  const existingKeysQuery = await db
    .select(Object.fromEntries(keyFields.map((field) => [field, table[field]])))
    .from(table);

  const existingKeys = new Set(
    existingKeysQuery.map((record) => createUniqueKey(record, keyFields)),
  );

  return processedData.filter((record) => {
    const identifier = createUniqueKey(record, keyFields);
    return !existingKeys.has(identifier);
  });
};

const insertRecordBatches = async <T extends PgTable>(
  table: T,
  newRecords: Record<string, any>[],
): Promise<{ totalInserted: number; duration: number }> => {
  let totalInserted = 0;
  const start = performance.now();

  for (let i = 0; i < newRecords.length; i += BATCH_SIZE) {
    const batch = newRecords.slice(i, i + BATCH_SIZE);
    const inserted = await db.insert(table).values(batch).returning();
    totalInserted += inserted.length;
    console.log(
      `Inserted batch of ${inserted.length} records. Total: ${totalInserted}`,
    );
  }

  const end = performance.now();
  const duration = Math.round(end - start);

  return { totalInserted, duration };
};

export const updater = async <T extends PgTable>({
  table,
  zipFileName,
  zipUrl,
  keyFields,
}: UpdaterConfig<T>): Promise<UpdaterResult> => {
  try {
    // Download and prepare file
    const destinationPath = await downloadAndPrepareFile(zipUrl);

    // Validate file checksum
    const { checksum, isNewOrChanged } = await validateFileChecksum(
      zipFileName,
      destinationPath,
    );

    // If file is not changed, return early
    if (!isNewOrChanged) {
      return {
        recordsProcessed: 0,
        message: `File has not been changed since the last update. Checksum: ${checksum}`,
        timestamp: new Date().toISOString(),
      };
    }

    // Process CSV
    const processedData = await processCSV(destinationPath);

    // Find new records
    const newRecords = await findNewRecords(table, processedData, keyFields);

    // Return early if no new records
    if (newRecords.length === 0) {
      return {
        recordsProcessed: 0,
        message:
          "No new data to insert. The provided data matches existing records.",
        timestamp: new Date().toISOString(),
      };
    }

    // Insert records in batches
    const { totalInserted, duration } = await insertRecordBatches(
      table,
      newRecords,
    );

    // Cache the new checksum
    await cacheChecksum(zipFileName, checksum);

    return {
      recordsProcessed: totalInserted,
      message: `${totalInserted} record(s) inserted in ${duration}ms`,
      timestamp: new Date().toISOString(),
    };
  } catch (error) {
    console.error("Error in updater:", error);
    return {
      recordsProcessed: 0,
      message: `Error during update: ${error instanceof Error ? error.message : "Unknown error"}`,
      timestamp: new Date().toISOString(),
    };
  }
};
